src/internal/m365/graph/concurrency_middleware.go

```diff
@@ -20,34 +20,41 @@ import (
 // "how many calls at one time"
 // ---------------------------------------------------------------------------
 
-// concurrencyLimiter middleware limits the number of concurrent requests to graph API
-type concurrencyLimiter struct {
+// concurrencyLimiterMiddleware middleware limits the number of concurrent requests to graph API
+type concurrencyLimiterMiddleware struct {
 	semaphore chan struct{}
 }
 
 var (
-	once                  sync.Once
-	concurrencyLim        *concurrencyLimiter
-	maxConcurrentRequests = 4
+	once                                sync.Once
+	concurrencyLimitMiddlewareSingleton = &concurrencyLimiterMiddleware{}
+	maxConcurrentRequests               = 4
 )
 
-func generateConcurrencyLimiter(capacity int) *concurrencyLimiter {
+func generateConcurrencyLimiter(capacity int) *concurrencyLimiterMiddleware {
 	if capacity < 1 || capacity > maxConcurrentRequests {
 		capacity = maxConcurrentRequests
 	}
 
-	return &concurrencyLimiter{
+	return &concurrencyLimiterMiddleware{
 		semaphore: make(chan struct{}, capacity),
 	}
 }
 
 func InitializeConcurrencyLimiter(capacity int) {
 	once.Do(func() {
-		concurrencyLim = generateConcurrencyLimiter(capacity)
+		concurrencyLimitMiddlewareSingleton.semaphore = generateConcurrencyLimiter(capacity).semaphore
 	})
 }
 
-func (cl *concurrencyLimiter) Intercept(
+func TurnOffConcurrencyLimiter(ctx context.Context) {
```

Comment:
Having two functions modify the `concurrencyLimitMiddlewareSingleton` feels wrong. It gives reader an illusion that they can toggle the singleton state during a run, from on to off & vice versa. 

Alternate proposal: What if we supply an `enabled` flag to `InitializeConcurrencyLimiter`. That way the singleton will be only initialized once, based on flag state?